game-core/src/main/java/games/strategy/ui/AutoCompletion.java

```diff
@@ -0,0 +1,232 @@
+package games.strategy.ui;
+
+import java.awt.event.ActionEvent;
+import java.awt.event.ActionListener;
+import java.awt.event.FocusAdapter;
+import java.awt.event.FocusEvent;
+import java.awt.event.FocusListener;
+import java.awt.event.KeyAdapter;
+import java.awt.event.KeyEvent;
+import java.awt.event.KeyListener;
+import java.beans.PropertyChangeEvent;
+import java.beans.PropertyChangeListener;
+
+import javax.swing.ComboBoxEditor;
+import javax.swing.ComboBoxModel;
+import javax.swing.JComboBox;
+import javax.swing.text.AttributeSet;
+import javax.swing.text.BadLocationException;
+import javax.swing.text.JTextComponent;
+import javax.swing.text.PlainDocument;
+
+/**
+ * From http://www.orbital-computer.de/JComboBox. Originally released into the Public Domain
```

Comment:
This is just a drop-in copy? Any thoughts if we should/can make a cleaner split between our triple custom UI code vs dropped-in code.

Reply:
Wasn't really sure where else to put it.  With 3cee3b5fc, I moved it to the `swinglib` package and made it package-private to hide it from general client usage.